virtcontainers/api.go

```diff
@@ -929,3 +929,56 @@ func ListRoutes(ctx context.Context, sandboxID string) ([]*vcTypes.Route, error)
 
 	return s.ListRoutes()
 }
+
+// CleanupContaienr is used by shimv2 to stop and delete a container exclusively, once there is no container
+// in the sandbox left, do stop the sandbox and delete it. Those serial operations will be done exclusively by
+// locking the sandbox.
+func CleanupContainer(ctx context.Context, sandboxID, containerID string, force bool) error {
+	span, ctx := trace(ctx, "CleanupContainer")
+	defer span.Finish()
+
+	if sandboxID == "" {
+		return vcTypes.ErrNeedSandboxID
+	}
+
+	if containerID == "" {
+		return vcTypes.ErrNeedContainerID
+	}
+
+	lockFile, err := rwLockSandbox(ctx, sandboxID)
+	if err != nil {
+		return err
+	}
+	defer unlockSandbox(ctx, sandboxID, lockFile)
+
+	s, err := fetchSandbox(ctx, sandboxID)
+	if err != nil {
+		return err
+	}
+
+	defer s.Release()
+
+	_, err = s.StopContainer(containerID, true)
+	if err != nil && !force {
+		return err
+	}
+
+	_, err = s.DeleteContainer(containerID)
+	if err != nil && !force {
+		return err
+	}
+
+	if len(s.GetAllContainers()) > 0 {
+		return nil
+	}
+
+	if err = s.Stop(true); err != nil && !force {
```

Comment:
Since `force` is passed as parameter, should use it here? Or we can just remove the `force` parameter and always enable `force` cleanup.

Reply:
Hi @bergwolf , Yes, you are right. I had change it using the parameter 'force' instead of 'true'.